<launch>
    <!-- Live or reading a rosbag file (offline) -->
    <arg name="offline" default="true" />
    <!-- Which semantic data file should be loaded? -->
    <arg name="env_database" default="ulms_seq03_jfk_e04_ofc.json" />

    <!-- Check to see if loads ORB-SLAM's TF values or not -->
    <param name="use_sim_time" value="$(arg offline)" />

    <!-- Main node -->
    <node name="orb_slam3" pkg="orb_slam3_ros" type="ros_stereo_inertial" output="screen">
        <!-- Remap to prepare the raw images -->
        <remap from="/camera/right/image_raw"   to="/camera/infra1/image_rect_raw"/>
        <remap from="/camera/left/image_raw"    to="/camera/infra2/image_rect_raw"/>

        <!-- Remap to prepare the inertial data -->
        <remap from="/imu"                      to="/camera/imu"/>

        <!-- Parameters for loading real semantic data from the environment -->
        <param name="env_file" type="string" value="$(find orb_slam3_ros)/config/Environments/$(arg env_database)"/>

        <!-- Parameters for loading setting files -->
        <param name="voc_file"      type="string" value="$(find orb_slam3_ros)/orb_slam3/Vocabulary/ORBvoc.txt.bin"/>
        <param name="settings_file" type="string" value="$(find orb_slam3_ros)/config/Stereo-Inertial/UniLu_RealSense_D435i_640.yaml"/>

        <!-- Parameters for ROS -->
        <param name="roll" type="double" value="0" />
        <param name="yaw" type="double" value="1.5697" />
        <param name="pitch" type="double" value="-1.5697" />
        <param name="map_frame_id"    type="string"   value="map" />
        <param name="imu_frame_id"      type="string"   value="imu" />
        <param name="enable_pangolin"   type="bool"     value="false" />
        <param name="world_frame_id"    type="string"   value="world" />
        <param name="cam_frame_id"      type="string"   value="camera" />
    </node>

    <!-- Transformation parameters -->
    <node pkg="tf" type="static_transform_publisher" name="wall_to_room" args="0 -3 0 0 0 0 wall room 10" />
    <node pkg="tf" type="static_transform_publisher" name="world_to_wall" args="0 -3 0 0 0 0 world wall 10" />
    <node pkg="tf" type="static_transform_publisher" name="camera_to_camera_optical" args="0 0 0 0 0 0 camera camera_color_optical_frame 10" />
    
    <!-- Visualization -->
    <node name="rviz" pkg="rviz" type="rviz" args="-d $(find orb_slam3_ros)/config/unilu_stereo_imu.rviz" output="screen" />
    
    <!-- Trajectory path -->
    <node pkg="hector_trajectory_server" type="hector_trajectory_server" name="trajectory_server_orb_slam3" output="screen" ns="orb_slam3_ros" >
        <param name="/source_frame_name" value="/imu" />
        <param name="/target_frame_name" value="/map" />
        <param name="/trajectory_update_rate" value="20.0" />
        <param name="/trajectory_publish_rate" value="20.0" />
    </node>
</launch>